drivers/rtc/rtc_ds3231.c

```diff
@@ -805,7 +808,9 @@ static int rtc_ds3231_init(const struct device *dev)
 	int err = 0;
 
 	const struct rtc_ds3231_conf *config = dev->config;
+#if defined(CONFIG_RTC_UPDATE) || defined(CONFIG_RTC_ALARM)
 	struct rtc_ds3231_data *data = dev->data;
+#endif
```

Comment:
The ifdefery can be reduced with `__maybe_unused` (although the position of it might need to be adjusted).

```suggestion
  __maybe_unused
 	struct rtc_ds3231_data * data = dev->data;
```

Reply:
Hey, good idea, at least for local variables. I'll do so and roll out a reworked patch set. For static functions I'm just unsure, if the compiler or linker optimization would skip out the symbols to reduce memory consumption and avoid dead code. Thus I would still ifdef-out them.